src/d/actor/d_a_obj_shelf.cpp

```diff
@@ -4,133 +4,288 @@
+        hold_event();
+        quake = false;
+        event = true;
+    }
+
+    if (quake) {
+        mode_vib_init();
+    } else if (event) {
+        mode_rot_init3();
+    }
 }
 
 /* 000003FC-00000424       .text mode_vib_init__Q210daObjShelf5Act_cFv */
 void daObjShelf::Act_c::mode_vib_init() {
-    /* Nonmatching */
+    mTimer = attr().mVibDuration;
+    mVibY = 0;
+    mVibX = 0;
+    mVibZ = 0;
+    mMode = 1;
 }
 
 /* 00000424-00000564       .text mode_vib__Q210daObjShelf5Act_cFv */
 void daObjShelf::Act_c::mode_vib() {
-    /* Nonmatching */
+    if (--mTimer <= 0) {
+        current.pos.y = home.pos.y;
+        shape_angle.x = home.angle.x;
+        shape_angle.z = home.angle.z;
+        mode_rot_init();
+    } else {
+        mVibY += attr().mVibYSpeed;
+        current.pos.y = home.pos.y + cM_ssin(mVibY) * attr().mVibYMagnitude;
+        mVibX += attr().mVibXSpeed;
+        mVibZ += attr().mVibZSpeed;
+        shape_angle.x = home.angle.x + (short)(cM_ssin(mVibX) * attr().mVibXMagnitude);
+        shape_angle.z = home.angle.z + (short)(cM_ssin(mVibZ) * attr().mVibZMagnitude);
+    }
 }
 
 /* 00000564-000005A8       .text mode_rot_init__Q210daObjShelf5Act_cFv */
 void daObjShelf::Act_c::mode_rot_init() {
-    /* Nonmatching */
+    mCurBounce = attr().mBounceNum + 1;
+    mRotSpeed = 0.0f;
+    mTimer = attr().mRotWaitDuration;
+    mTargetAngle = 0x4000;
+    m2e7 = false;
+    mMode = 2;
 }
 
 /* 000005A8-000005E4       .text mode_rot_init2__Q210daObjShelf5Act_cFv */
 void daObjShelf::Act_c::mode_rot_init2() {
-    /* Nonmatching */
+    mCurBounce = attr().mBounceNum + 1;
+    mRotSpeed = attr().mInitSpeed2;
+    mTimer = 0;
+    mTargetAngle = 0x4000;
+    m2e7 = false;
+    mMode = 2;
 }
 
 /* 000005E4-00000620       .text mode_rot_init3__Q210daObjShelf5Act_cFv */
 void daObjShelf::Act_c::mode_rot_init3() {
-    /* Nonmatching */
+    mCurBounce = attr().mBounceNum + 1;
+    mRotSpeed = attr().mInitSpeed3;
+    mTimer = 0;
+    mTargetAngle = 0;
+    m2e7 = true;
+    mMode = 2;
 }
 
 /* 00000620-000007AC       .text mode_rot__Q210daObjShelf5Act_cFv */
 void daObjShelf::Act_c::mode_rot() {
-    /* Nonmatching */
+    if (mTimer > 0) {
+        mTimer--;
+        return;
+    }
+    mRotSpeed += attr().mRotAccel;
+    mRotSpeed -= mRotSpeed * attr().mRotDecay;
+    shape_angle.x += (short)(int)mRotSpeed;
+    if (shape_angle.x > mTargetAngle) {
+        if (--mCurBounce <= 0) {
+            shape_angle.x = mTargetAngle;
+            if (m2e7) {
+                mode_wait_init();
+            } else {
+                mode_fell_init();
+            }
+        } else {
+            // Bounce
+            shape_angle.x = mTargetAngle - (int)((short)(int)(shape_angle.x - mTargetAngle) * attr().mBounceFactor);
+            mRotSpeed = mRotSpeed * -attr().mBounceFactor;
+            if (mCurBounce == attr().mBounceNum) {
+                fopAcM_seStart(this, 0x696d, 0);
```

Comment:
```suggestion
                fopAcM_seStart(this, JA_SE_OBJ_RACK_BREAK, 0);
```
JAZelAudio_SE.h has an enum of all the official sound effect names